search_similar_entities/facetedSearchEndpoint.rb

```diff
@@ -0,0 +1,161 @@
+  return testEntity['hierarchicalAttr1'].select { |a| a.start_with?(level1value + '|' + level2value) }.uniq
+end
+
+# Utility method that adds n and k elements to a hash
+def addNKElementsToHash(theHash, nObj, kObj)
+  unless nObj.nil? or nObj.empty?
+    theHash['n'] = nObj
+  end
+  unless kObj.nil? or kObj.empty?
+    theHash['k'] = kObj
+  end
+  return theHash
+end
+  
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget is selected for a given value
+def getSimpleRefWidgetDataStruct(widgetName,xPath,logic,theValue)
+  
+  theValueHash = {"n" => theValue}
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => [theValueHash]}
+  unless xPath.empty?
+    tmpHash['xPath'] = xPath
+  end
+  return tmpHash
+  
+end 
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget (that has hierarchical values) is selected for a given value
+def getComplexRefWidgetDataStruct(widgetName,xPath,logic,theValue,separator)
+  
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => theValue}
+  unless xPath.empty?
+    tmpHash['xPath'] = xPath
+  end
+  unless separator.empty?
+    tmpHash['separator'] = separator
+  end
+  return tmpHash
+  
+end 
+
+# Utility method to return the complex data structure that goes in the queryURL for a
+# nested refinement widget
+# theLabels has a number of labels separated by separator
+# starting from backwards, it adds the label, and the full path to the label
+def getComplexRefWidgetDataStructure(theLabels, separator)
+  labelArray = theLabels.split('|')
+  arrayOfHashes = []
+  until labelArray.empty? do
+    allLabels = labelArray.join('|')
+    h = {}
+    h = addNKElementsToHash(h,labelArray.last, allLabels)
+    arrayOfHashes << h
+    labelArray.pop # Remove the last label
+  end
+  # Now iterate over all the building blocks we've got so far, nesting each element as an "s" in the next hast
+  for i in 0 .. arrayOfHashes.count - 2
+    arrayOfHashes[i+1]['s'] = [arrayOfHashes[i]]
+  end
+  return arrayOfHashes.last
+end
+
+## END Utility Methods
+
+# Now the real thing begins. Customise the search criteria as required
+
+# This query selects the testEntity entity to be used as a starting point
+thisTestEntity = entity_type('testEntity').where(field('id').is(thisId)).first
+
+# Check that we have the basic attributes, otherwise return an basic query
+
+if thisTestEntity['simpleAttr1'].nil? or thisTestEntity['simpleAttr1'].empty?
+    output = URI.encode('../search?q=entityType:testEntity')
+else
+# Start building the query string: Get the list of testEntity entities that match the simpleAttr1 value of this testEntity (and are not this testEntity)
+
+  similarTestEntitiesQueryString = '../search?q=simpleAttr1:\''+thisTestEntity['simpleAttr1'].first  + '\' AND entityType:testEntity AND NOT(id:'+ thisTestEntity['id'].first+')'
```

Comment:
I generally recommend erb (or some other templating) instead of string concatenation.  To recode the query using erb and `:to_query` :

```
query_attributes = {:attr_one =>  thisTestEntity['simpleAttr1'].first, :attr_id =>   thisTestEntity['id'].first }
query_template = %{entityType:testEntity AND simpleAttr1:<%= query_attributes[:attr_one] %> AND NOT(id:<%= query_attributes[:attr_id] %>)}

similarTestEntitiesQueryString = {q: ERB.new(query_template).result(binding) }.to_query
```